src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts

```diff
@@ -0,0 +1,95 @@
+import { pullAt, escapeRegExp } from 'lodash';
+import { OptimizerPlugin } from '../OptimizerPlugin';
+import { Package } from '../../processor';
+import {
+  ExportableAssignmentRule,
+  ExportableCaretValueRule,
+  ExportableInstance,
+  ExportableSdRule
+} from '../../exportable';
+
+export default {
+  name: 'construct_inline_instance',
+  description:
+    'Construct inline instances from groups of rules in a contained resource or a Bundle',
+
+  optimize(pkg: Package): void {
+    const inlineInstances: ExportableInstance[] = [];
+    [...pkg.instances, ...pkg.profiles].forEach(resource => {
+      const ruleType =
+        resource instanceof ExportableInstance
+          ? ExportableAssignmentRule
+          : ExportableCaretValueRule;
+      // First get all possible paths for which child elements should be extracted onto an inline instance
+      const basePaths = resource.rules
+        .filter(
+          rule =>
+            rule instanceof ruleType &&
+            (/^contained(\[\d+\])?\.resourceType$/.test(getRulePath(rule)) ||
+              /^entry(\[\d+\])?\.resource\.resourceType$/.test(getRulePath(rule)))
```

Comment:
It seems possible it's something we would add, and I'm a bit worried we would forget to specify that it runs after this one, since inline instances are such a specific case that we might not notice it things like this.